### ISSUE TITLE
Error is not displayed when a script crashes in gpfdist output transform

### ISSUE THREAD

**Issue by oSkrobuk** [greenplum-db/gpdb#9497]
### Greenplum version or build
PostgreSQL 8.3.23 (Greenplum Database 5.23.1)

### OS version and uname -a
Distributor ID: Ubuntu
Description: Ubuntu 14.04.1 LTS
Release : 14.04
Codename: trusty
Linux ******* 4.4.114-50 #1 SMP Fri Feb 2 10:53:13 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### autoconf options used ( config.status --config )
BINDIR = /usr/local/gpdb/bin
DOCDIR = /usr/local/gpdb/doc/postgresql
HTMLDIR = /usr/local/gpdb/doc/postgresql
INCLUDEDIR = /usr/local/gpdb/include
PKGINCLUDEDIR = /usr/local/gpdb/include/postgresql
INCLUDEDIR-SERVER = /usr/local/gpdb/include/postgresql/server
LIBDIR = /usr/local/gpdb/lib
PKGLIBDIR = /usr/local/gpdb/lib/postgresql
LOCALEDIR = /usr/local/gpdb/share/locale
MANDIR = /usr/local/gpdb/share/man
SHAREDIR = /usr/local/gpdb/share/postgresql
SYSCONFDIR = /usr/local/gpdb/etc/postgresql
PGXS = /usr/local/gpdb/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/local/gpdb/include
CFLAGS = -O3 -std=gnu99 -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -I/usr/local/gpdb/include
CFLAGS_SL = -fpic
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/usr/local/gpdb/lib',--enable-new-dtags -L/usr/local/gpdb/lib
LDFLAGS_SL =
LIBS = -lpgport -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lxml2 -lrt -lcrypt -ldl -lm -L/usr/local/gpdb/lib
VERSION = PostgreSQL 8.3.23

### Expected behavior
Forward an error to GP when it occurs in the work of the output transformation script

### Step to reproduce the behavior
Transformation config:
```
TRANSFORMATIONS:
    write:
        TYPE: output
        COMMAND: /usr/bin/env /home/user/gpfdist/error.sh %filename%
    read:
        TYPE: input
        COMMAND: /usr/bin/env /home/user/gpfdist/error.sh %filename%
```
error.sh:
```
#!/bin/bash

error_command
```
Create external tables:
```
CREATE WRITABLE EXTERNAL TABLE public.ext_write
(
  column VARCHAR
)
LOCATION ('gpfdist://hostname:8090/filename#transform=write')
FORMAT 'csv' 
ENCODING 'UTF8';

CREATE EXTERNAL TABLE public.ext_read
(
  column VARCHAR
)
LOCATION ('gpfdist://hostname:8090/filename#transform=read')
FORMAT 'csv' 
ENCODING 'UTF8';
```
Run gpfdist:
```
gpfdist -p 8090 -d /home/user/gpfdist -c /home/user/gpfdist/transformation.yaml -V
```
Execute queries:
```
INSERT INTO public.ext_write
SELECT 'text';
--executing without error
SELECT * FROM public.ext_read;
--[22000] ERROR: gpfdist error - /home/user/gpfdist/error.sh: line 3: error_command: command not found
```
When set STDERR value CONSOLE in the transform configuration for output, I see an error in the gpfdist log
```
TRANSFORMATIONS:
    write:
        TYPE: output
        COMMAND: /usr/bin/env /home/user/gpfdist/error.sh %filename%
        STDERR: console
    read:
        TYPE: input
        COMMAND: /usr/bin/env /home/user/gpfdist/error.sh %filename%
```
gpfdist log:
```
2020-01-23 17:40:53 407193 INFO [0:3:0:7] Content-Length:0
 write:
  TYPE: output
  COMMAND: /usr/bin/env /home/user/gpfdist/error.sh %filename%
  STDERR: console
  CONTENTS: data
2020-01-23 17:40:53 407193 INFO [0:3:56:7] transform: write
2020-01-23 17:40:53 407193 INFO [0:3:56:7] joined session (/home/user/gpfdist/file_pattern, 1575890863-0001326025.0.0.0)
2020-01-23 17:40:53 407193 INFO active segids in session:
2020-01-23 17:40:53 407193 INFO [1:3:56:7] IPv6_addres:7375 POST /file_pattern - OK
2020-01-23 17:40:53 407193 INFO [1:3:56:7] request end
2020-01-23 17:40:53 407193 INFO [1:3:56:7] detach segment request from session
2020-01-23 17:40:53 407193 INFO free session 1575890863-0001326025.0.0.0:/home/user/gpfdist/file_pattern
/home/user/gpfdist/error.sh: line 3: error_command: command not found
close_subprocess: done: why = 1, exit status = 127
[warn] event_del: event has no event_base set.
2020-01-23 17:40:53 407193 INFO [1:3:56:7] successfully shutdown socket
2020-01-23 17:40:53 407193 INFO [1:3:56:7] peer closed after gpfdist shutdown
2020-01-23 17:40:53 407193 INFO [1:3:56:7] unsent bytes: 0 (-1 means not supported)
2020-01-23 17:40:53 407193 INFO [1:3:56:7] successfully closed socket
```
Please, help. A very critical issue for me

**Comment by sokolval** [comment]
Hello guys, I have a similar issue. 
Is there any fix for this?

**Comment by lij55** [comment]
It looks the errer message is clear. 

> error_command: command not found

so, what happen if you run `error_command` directly?

**Comment by lij55** [comment]
ohh, I misunder the issue. Please ignore my last message. Sorry for that.

to cleary, you need this error message "error_command: command not found" going to the log of Greenplum under pg_log. Is it correct?

**Comment by sokolval** [comment]
Hey, Jasper, thanks for getting back with that!
What is needed is to have same behavior for output transform as it is for input: if there is error it should throw exception and break, which is not the case now.

**Comment by zhaorui20005** [comment]
This is a bug of *transform* in gpfdist. After reading the code of gpfdist, I found that it is caused by the transform writing function "write_subprocess" which does not handle the return value of "apr_file_write". I will make a pr later to fix this.

**Comment by wfnuser** [comment]
Any update? I have met this problem too. @zhaorui20005

**Comment by zhaorui20005** [comment]
You can simply modify the function like this to fix this issue:
```static ssize_t
write_subprocess(gfile_t *fd, void *ptr, size_t size)
{
	apr_size_t      nbytes = size;
	apr_status_t    rv;

	rv = apr_file_write(fd->transform->proc.in, ptr, &nbytes);
	if(rv == APR_SUCCESS)
		return nbytes;
	else
		return -1;
}
```
But it is not stable. There is something wrong with the client codes.

**Comment by wfnuser** [comment]
> 

Yep. I have tried that... But there are more problems. I can just get 500 from gpfdist. But not detailed error messages.
I figured that the gpfdist has sent error message, but the client won't read the response body in time and broke.
It will only received the SIGPIPE error from gpfdist when it do `proto0_write`... But the error message is sent back only in `curl_fopen`.

**Comment by wfnuser** [comment]
The problem might be when QE curl_fopen... gpfdist send error message back, but the QE only check response header but not the response body with error message.
How can I solve that. I found that there are `still_running` variable in fill_buffer. But in write path, we just use `curl_easy_perform`. How can I get response body in write path.

**Comment by wfnuser** [comment]
I think the real problem is that when gpfdist received the open request.
It will spawn a subprocess to do the transform job by calling subprocess_open, but finally it will fail in this case. However, the error happened in the subprocess won't be written to errfile in time. And when we do apr_file_read, nothing will be read from the errfile. That's why we can't get error message from server.
@zhaorui20005

**Comment by zhaorui20005** [comment]
@wfnuser A new pull request has been open for this issue. Feel free to add comments. https://github.com/greenplum-db/gpdb/pull/13442

**Comment by wfnuser** [comment]
> 

Cool. I have left some comments.

**Comment by ashwinstar** [comment]
@huiliang-liu this seems to be heavily complained about defect in product. I see the associated PR opened and then due to no further traction had to be closed. Do we have plans to drive attention on resolving this?

**Comment by huiliang-liu** [comment]
@ashwinstar As I known, it is not urgent now. The previous PR is not very good to fix this issue too. We can drive attention on this issue later. Thank you for your notification.
